crates/cuda-common/src/memory_manager/tests.rs

```diff
@@ -0,0 +1,505 @@
+    // +11 needs defrag of the 6 + allocate (11-X-6) new pages
+    let initial_pages = 12; // X = 12 - 11 = 1
+
+    let (pool, page_size, ptr_1, ptr_4, ptr_11) = run_doc_scenario(initial_pages);
+    let stream_id = current_stream_id().unwrap();
+
+    // Memory usage: need 11 more pages but only have 6+1=7 free
+    // So allocate 11-7=4 new pages → 12 + 4 = 16 total
+    assert_eq!(
+        pool.memory_usage(),
+        16 * page_size,
+        "Case D: 4 new pages allocated"
+    );
+
+    // +4 at VA start (takes from freed 10 since X < 4)
+    assert_eq!(
+        ptr_4 as usize, pool.roots[0] as usize,
+        "Case D: 4 at VA start"
+    );
+
+    // Cleanup
+    let mut pool = pool;
+    pool.free_internal(ptr_1, stream_id).unwrap();
+    pool.free_internal(ptr_4, stream_id).unwrap();
+    pool.free_internal(ptr_11, stream_id).unwrap();
+}
+
+// ============================================================================
+// Cross-thread: Thread A frees, Thread B should be able to reuse after sync
+// ============================================================================
+#[test]
+fn test_cross_thread_handoff() {
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .worker_threads(2)
+        .max_blocking_threads(2)
+        .enable_all()
+        .build()
+        .unwrap();
+
+    runtime.block_on(async {
+        let barrier = Arc::new(Barrier::new(2));
+        // Store address as usize (which is Send + Sync)
+        let addr_holder = Arc::new(std::sync::Mutex::new(0usize));
+
+        let barrier1 = barrier.clone();
+        let addr_holder1 = addr_holder.clone();
+
+        // Thread A: allocate and free
+        let handle_a = tokio::task::spawn(async move {
+            tokio::task::spawn_blocking(move || {
+                let len = (200 << 20) / size_of::<u8>(); // 200 MB
+                let buf = DeviceBuffer::<u8>::with_capacity(len);
+                *addr_holder1.lock().unwrap() = buf.as_raw_ptr() as usize;
+                // buf drops here, freeing memory
+            })
+            .await
+            .expect("task A panicked");
+
+            barrier1.wait().await;
+        });
+
+        let barrier2 = barrier.clone();
+        let addr_holder2 = addr_holder.clone();
+
+        // Thread B: wait for A, then allocate same size
+        let handle_b = tokio::task::spawn(async move {
+            barrier2.wait().await;
+
+            tokio::task::spawn_blocking(move || {
+                let len = (200 << 20) / size_of::<u8>();
+                let buf = DeviceBuffer::<u8>::with_capacity(len);
+
+                // Check if we got the same address (depends on event sync timing)
+                let original_addr = *addr_holder2.lock().unwrap();
+                let new_addr = buf.as_raw_ptr() as usize;
+                if new_addr == original_addr {
+                    println!("Cross-thread: reused same address (event synced)");
```

Comment:
if you only print, this test will not automatically catch anything: can you force exactly one scenario to occur and assert! it?